### PR TITLE
vdk-heartbeat: Make vdkcli_oauth2_uri and control_api_url optional

### PR DESCRIPTION
**Pull request by YanaZhivkova**
If we want to test VDK's configuration and default options, we want
to be able to set those params values in the default vdkcli options
and run vdk-heartbeat.

Make vdkcli_oauth2_uri and control_api_url params optional and don't
pass values if they are not set in the config.ini - thus whatever is
set as default in VDK's configuration will be used.

Tested by: Run locally built vdk-heartbeat in an environment with
properly configured defaults with a config.ini looking like this:
[DEFAULT]
VDKCLI_OAUTH2_REFRESH_TOKEN=***

JOB_RUN_TEST_MODULE_NAME=vdk.internal.heartbeat.simple_run_test
JOB_RUN_TEST_CLASS_NAME=SimpleRunTest
DATAJOB_DIRECTORY_NAME=simple
VDK_COMMAND_NAME=vdk

Signed-off-by: Yana Zhivkova <yzhivkova@vmware.com>

### CODE REVIEW

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py [hunk 1]
```diff
@@ -102,9 +104,10 @@ def __init__(self, config_file=None):
         )
 
         # The Control Service API URL (http://url/data-jobs) without data-jobs suffix
-        self.control_api_url = self._get_atleast_one_value(
-            "CONTROL_API_URL", "VDK_HEARTBEAT_CONTROL_SERVICE_URL"
-        )
+        # If no value is found, argument will not be passed and default will be used
+        self.control_api_url = self.get_value(
+            "CONTROL_API_URL", is_required=False
+        ) or self.get_value("VDK_HEARTBEAT_CONTROL_SERVICE_URL", is_required=False)
```
Comment (tozka) [inline projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py:110]: minor comment: Wouldn't be simpler to add `is_required` argument to `_get_atleast_one_value` ?

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py [hunk 2]
```diff
@@ -23,10 +23,18 @@ class JobController:
 
     def __init__(self, config: Config):
         self.config = config
+        control_api_url = (
```
Comment (tozka) [inline projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py:26]: control_api_url_message seems more explicit and clear name for the variable here I think

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py [hunk 3]
```diff
@@ -23,10 +23,18 @@ class JobController:
...
+            config.control_api_url
+            or "Not set (default from VDK's configuration will be used)"
+        )
+        auth_endpoint = (
```
Comment (tozka) [inline projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py:30]: `auth_endpoint_message` ?

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py [hunk 4]
```diff
@@ -46,102 +54,174 @@ def _execute(self, command):
...
+        else:
+            self._execute(
+                [
+                    "login",
+                    "-a",
+                    f"{self.config.vdkcli_api_refresh_token}",
+                    "-t",
+                    "api-token",
+                ]
+            )
```
Comment (tozka) [inline projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py:78]:
```

    def __control_api_url_arg(self):
        if self.config.vdkcli_oauth2_uri:
            return ["-u", f"{self.config.vdkcli_oauth2_uri}"]
        else 
            return []

    @LogDecorator(log)
    def login(self):
        self._execute(["login"] + ["-t", "api-token"] + self.__control_api_url_arguments())
```

Perhaps something akin to above would be slightly simpler and clearer. Here in those arrays people would be playing a bit [Spot the difference](https://en.wikipedia.org/wiki/Spot_the_difference) :)

**Review by tozka — COMMENTED** [review]
Looks good to me. I have only a few minor readability comments.

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py (hunk 1 above)
Comment (YanaZhivkova) [reply projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py]: Yes, good point

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py (hunk 3 above)
Comment (YanaZhivkova) [reply projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py]: Done.

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py (hunk 4 above)
Comment (YanaZhivkova) [reply projects/vdk-heartbeat/src/vdk/internal/heartbeat/job_controller.py]: Yes, thank you! I spent some time trying to find a way to do that, but everything I could think of, was awful :D

File: projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py [hunk 5]
```diff
@@ -224,7 +227,7 @@ def _get_atleast_one_value(self, key1: str, key2: str) -> Optional[str]:
                 f"Countermeasures: Set as required configuration either as environment variable or in the config file."
             )
 
-        return value1 if value1 is not None else value2
+        return value1 or value2 or False
```
Comment (tozka) [inline projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py:230]: No, this won't work. Don't you have some unit test failing? If not consider adding unit tests that return empty string for eample

If value1 and value2 are a boolean (with value False) or are a empty string or empty array   

 they all evaluate to "false" in if check which would break the logic.

See https://docs.python.org/2.4/lib/truth.html

The previous logic seemed fine. Why changing it?

Comment (YanaZhivkova) [reply projects/vdk-heartbeat/src/vdk/internal/heartbeat/config.py]: Ok, I added some tests. I did not find a way to set value1/value2 to boolean values, they are always parsed as strings. But the empty string was a valid case.

**Review by tozka — APPROVED** [review]
